### PR TITLE
Added `empircal_sinkhorn_divergence` function

### PR DESCRIPTION
**Pull request by davibarreira**

### CODE REVIEW

**Review by devmotion — CHANGES_REQUESTED** [review]
Thanks for the PR! Can you make the new function consistent with the existing ones? Ie,
- remove all keyword arguments
- remove all more detailed explanations and link to POT instead
- add a doctest example such that coverage stays at 100%

File: src/lib.jl [hunk 1]
```diff
@@ -244,6 +244,38 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
     return pot.sinkhorn2(μ, ν, PyCall.PyReverseDims(permutedims(C)), ε; kwargs...)
 end
 
+"""
+    empirical_sinkhorn_divergence(X_s, X_t, reg; kwargs...)
```
Comment (devmotion) [inline src/lib.jl:248]: We use other argument names that are more consistent with Julia notation (e.g. lowercase and no underscores if possible):
```suggestion
    empirical_sinkhorn_divergence(xsource, xtarget, ε; kwargs...)
```

File: src/lib.jl [hunk 2]
```diff
@@ -244,6 +244,38 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+
+Compute the Sinkhorn Divergence from empirical data, where
+`X_s` and `X_t` are arrays representing samples in the source domain and target domain, respectively,
+and  `reg` is the regularization term. `a` and `b` are optinal sample weights in the source
+and target domain, respectively.
```
Comment (devmotion) [inline src/lib.jl:253]: This should be updated accordingly. `a` and `b` should not be explained, they are keyword arguments.

File: src/lib.jl [hunk 3]
```diff
@@ -244,6 +244,38 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+
+This function is a wrapper of the function
+[`ot.bregman.empirical_sinkhorn_divergence`](https://pythonot.github.io/all.html#ot.emd) in the Python Optimal Transport
```
Comment (devmotion) [inline src/lib.jl:256]: The link is wrong.

File: src/lib.jl [hunk 4]
```diff
@@ -244,6 +244,38 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+package. Keyword arguments are listed in the documentation of the Python function.
+
+```jldoctest empirical_sinkhorn_divergence
```
Comment (devmotion) [inline src/lib.jl:261]:
```suggestion
# Examples

```jldoctest
```

File: src/lib.jl [hunk 5]
```diff
@@ -244,6 +244,38 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+julia> X_s = [1.0, 0.0, 3.0];
```
Comment (devmotion) [inline src/lib.jl:262]:
```suggestion
julia> xsource = [1.0, 0.0, 3.0];
```

File: src/lib.jl [hunk 6]
```diff
@@ -244,6 +244,38 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+
+julia> X_t = [2.0, 3.0, 4.0];
```
Comment (devmotion) [inline src/lib.jl:264]:
```suggestion
julia> xtarget = [2.0, 3.0, 4.0];
```

File: src/lib.jl [hunk 7]
```diff
@@ -244,6 +244,38 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+
+julia> round(empirical_sinkhorn_divergence(X_s,X_t,1); sigdigits=3)
+3.03
+```
+
+See also: [`sinkhorn2`](@ref)
+"""
+function empirical_sinkhorn_divergence(X_s, X_t, reg; kwargs...)
```
Comment (devmotion) [inline src/lib.jl:272]:
```suggestion
function empirical_sinkhorn_divergence(xsource, xtarget, ε; kwargs...)
```

File: src/lib.jl [hunk 8]
```diff
@@ -244,6 +244,38 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+    return pot.bregman.empirical_sinkhorn_divergence(
+        typeof(X_s) <: Vector ? reshape(X_s, :, 1) : X_s,
+        typeof(X_t) <: Vector ? reshape(X_t, :, 1) : X_t,
```
Comment (devmotion) [inline src/lib.jl:275]: This is incorrect for more general vector types. Does it not work automatically or with PyReverseDims?

File: src/lib.jl [hunk 9]
```diff
@@ -244,6 +244,38 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+        reg;
+        kwargs...,
+    )[1]
```
Comment (devmotion) [inline src/lib.jl:278]: We do not extract return values - e.g. users might set a keyword argument such that the logs are returned as well.
```suggestion
    )
```

File: src/lib.jl (hunk 8 above)
Comment (davibarreira) [reply src/lib.jl:275]: It does not work automatically, it requires vectors to be `n x 1`  matrices. I never used the PyReverseDims.

File: src/lib.jl [hunk 10]
```diff
@@ -244,6 +244,40 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+
+# Examples
+
+```jldoctest empirical_sinkhorn_divergence
+julia> xsource = [1.0, 0.0, 3.0];
+
+julia> xtarget = [2.0, 3.0, 4.0];
+
+julia> round(empirical_sinkhorn_divergence(X_s,X_t,1); sigdigits=3)
+j3.03
```
Comment (devmotion) [inline src/lib.jl:267]:
```suggestion
julia> round(empirical_sinkhorn_divergence(xsource, xtarget, 1); sigdigits=3)
3.03
```

File: src/lib.jl (hunk 8 above)
Comment (davibarreira) [reply src/lib.jl:275]: For what situation this doesn't work?

Comment (devmotion) [reply src/lib.jl:275]: If matrices are needed you can try
```suggestion
        reshape(xsource, Val(2)),
        reshape(xtarget, Val(2)),
```
This reshapes vectors (not only of type `Vector`) to matrices and keeps matrices as matrices.

File: src/lib.jl [hunk 11]
```diff
@@ -244,6 +244,40 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+```
+
+See also: [`sinkhorn2`](@ref)
+"""
+function empirical_sinkhorn_divergence(xsource, xtarget, ε; kwargs...)
+    return pot.bregman.empirical_sinkhorn_divergence(
+        typeof(X_s) <: Vector ? reshape(X_s, :, 1) : X_s,
+        typeof(X_t) <: Vector ? reshape(X_t, :, 1) : X_t,
+        reg;
```
Comment (devmotion) [inline src/lib.jl:276]:
```suggestion
        ε;
```

File: src/lib.jl (hunk 8 above)
Comment (devmotion) [reply src/lib.jl:275]: > For what situation this doesn't work?

If you use a vector that is not of type `Vector`.

File: src/lib.jl [hunk 12]
```diff
@@ -244,6 +244,36 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
     return pot.sinkhorn2(μ, ν, PyCall.PyReverseDims(permutedims(C)), ε; kwargs...)
 end
 
+"""
+    empirical_sinkhorn_divergence(xsource, xtarget, ε; kwargs...)
+
+Compute the Sinkhorn Divergence from empirical data, where
+`xsource` and `xtarget` are arrays representing samples in the source domain and target domain, respectively,
+and  `ε` is the regularization term.
```
Comment (devmotion) [inline src/lib.jl:252]:
```suggestion
Compute the Sinkhorn divergence from empirical data, where `xsource` and `xtarget` are
arrays representing samples in the source domain and target domain, respectively, and `ε`
is the regularization term.
```

File: src/lib.jl [hunk 13]
```diff
@@ -244,6 +244,36 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+
+This function is a wrapper of the function
+[`ot.bregman.empirical_sinkhorn_divergence`](https://pythonot.github.io/gen_modules/ot.bregman.html?highlight=sinkhorn%20divergence#ot.bregman.empirical_sinkhorn_divergence)
```
Comment (devmotion) [inline src/lib.jl:255]:
```suggestion
[`ot.bregman.empirical_sinkhorn_divergence`](https://pythonot.github.io/gen_modules/ot.bregman.html#ot.bregman.empirical_sinkhorn_divergence)
```

File: src/lib.jl [hunk 14]
```diff
@@ -244,6 +244,36 @@ function sinkhorn2(μ, ν, C, ε; kwargs...)
...
+in the Python Optimal Transport package. Keyword arguments are listed in the documentation of the Python function.
+
+# Examples
+
+```jldoctest empirical_sinkhorn_divergence
+julia> xsource = [1.0, 0.0, 3.0];
+
+julia> xtarget = [2.0, 3.0, 4.0];
+
+julia> round(empirical_sinkhorn_divergence(xsource,xtarget,1)[1]; sigdigits=3)
```
Comment (devmotion) [inline src/lib.jl:265]:
```suggestion
julia> round(empirical_sinkhorn_divergence(xsource, xtarget, 1)[1]; sigdigits=3)
```

**Review by devmotion — COMMENTED** [review]
Looks good, just some minor suggestions left. Can you update the version number as well?

**Comment by davibarreira** [comment]
I don't know what is this error with JuliaNightly, but it looks like it's not actually related to my PR.

**Comment by devmotion** [comment]
> I don't know what is this error with JuliaNightly, but it looks like it's not actually related to my PR.

No, they are known and caused by some deprecation in the Julia master branch that is not handled in PyCall yet (due to the deprecations the output of the doctests is different).

**Comment by devmotion** [comment]
Thanks for the PR, I'll make a new release right away!

**Pull request merged**